### PR TITLE
File system support and migration tool for file answers

### LINKED ISSUE #4251: File answers to file system
Framework for storing new file answers to file system and a way to transfer existing answers there as well.

### PR DESCRIPTION
**Pull request by pkukkon**
- resolves #4251

### CODE REVIEW

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/WorkspaceMaterialFieldAnswerController.java [hunk 1]
```diff
@@ -231,25 +241,13 @@ public WorkspaceMaterialFileFieldAnswerFile findWorkspaceMaterialFileFieldAnswer
   }
   
   public void deleteWorkspaceMaterialFileFieldAnswerFile(WorkspaceMaterialFileFieldAnswerFile fieldAnswerFile) {
+    if (fileAnswerUtils.isFileSystemStorageEnabled()) {
+      Long userEntityId = fieldAnswerFile.getFieldAnswer().getReply().getUserEntityId();
+      fileAnswerUtils.removeFileFromFileSystem(userEntityId, fieldAnswerFile.getFileId());
```
Comment (anttiviljakainen) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/WorkspaceMaterialFieldAnswerController.java]: Should this check if the file is even expected to be there? (it does silently skip it though if not)

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  
+  public boolean isFileSystemStorageEnabled() {
+    SystemSetting enabledSetting = systemSettingDAO.findByKey(SETTING_ENABLED);
+    return enabledSetting != null && Boolean.valueOf(enabledSetting.getValue());
+  }
+  
+  public byte[] getFileContent(Long userEntityId, String fileId) throws IOException {
+    if (userEntityId == null || fileId == null) {
+      logger.severe(String.format("Invalid fetch parameters: file %s of user %d", fileId, userEntityId));
+      return new byte[0];
```
Comment (anttiviljakainen) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java]: Might warrant throwing exceptions like https://docs.oracle.com/javase/7/docs/api/java/lang/IllegalArgumentException.html

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    }
+    SystemSetting fileFolderSetting = systemSettingDAO.findByKey(SETTING_FILEFOLDER);
+    if (fileFolderSetting != null && !StringUtils.isBlank(fileFolderSetting.getValue())) {
+      java.io.File answerFile = Paths.get(fileFolderSetting.getValue(), userEntityId.toString(), fileId).toFile();
+      if (answerFile.exists()) {
+        return Files.readAllBytes(answerFile.toPath());
+      }
+    }
+    logger.warning(String.format("Failed to retrieve file %s of user %d", fileId, userEntityId));
+    return new byte[0];
```
Comment (anttiviljakainen) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java]: Should this be exception-worth too?

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  }
+  
+  public void storeFileToFileSystem(Long userEntityId, String fileId, byte[] content) throws IOException {
+    if (userEntityId == null || fileId == null || content == null) {
+      logger.severe(String.format("Invalid store parameters or no file content: file %s of user %d", fileId, userEntityId));
+      return;
```
Comment (anttiviljakainen) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java]: Probably needs to be exception here to indicate the file couldn't be saved at all.

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    catch (Exception e) {
+      logger.log(Level.SEVERE, String.format("Failed to create user file %s", answerFile.getPath()), e);
+      throw new IOException("Failed to store file data", e);
+    }
+  }
+  
+  public void removeFileFromFileSystem(Long userEntityId, String fileId) {
+    if (userEntityId == null || fileId == null) {
+      logger.severe(String.format("Invalid remove parameters: file %s of user %d", fileId, userEntityId));
+      return;
```
Comment (anttiviljakainen) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java]: Illegal arguments?

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceRESTService.java [hunk 6]
```diff
@@ -1604,7 +1608,17 @@ public Response getFileAnswer(@PathParam("FILEID") String fileId) {
       }
     }
     
-    return Response.ok(answerFile.getContent())
+    byte[] content = answerFile.getContent();
+    if (content == null) {
+      try {
+        content = fileAnswerUtils.getFileContent(workspaceMaterialReply.getUserEntityId(), answerFile.getFileId());
```
Comment (anttiviljakainen) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceRESTService.java]: getFileContent currently may return byte[0] if not changed - should this be then status(NOT_FOUND)?

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceRESTService.java [hunk 7]
```diff
@@ -1707,7 +1721,11 @@ public void write(OutputStream out) throws IOException {
             
             ZipEntry ze = new ZipEntry(fileName);
             zout.putNextEntry(ze);
-            InputStream input = new ByteArrayInputStream(file.getContent());
+            byte[] content = file.getContent();
+            if (content == null) {
+              content = fileAnswerUtils.getFileContent(workspaceMaterialReply.getUserEntityId(), file.getFileId());
```
Comment (anttiviljakainen) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceRESTService.java]: getFileContent currently may return byte[0] if not changed - should this be checked?

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceSystemRESTService.java [hunk 8]
```diff
@@ -66,10 +75,56 @@
...
+          }
+          currentId = answerId;
+        }
+      }
+      catch (IOException e) {
+        logger.log(Level.SEVERE, String.format("Failed to relocate WorkspaceMaterialFileFieldAnswer %d", currentId), e);
+        return Response.status(Status.INTERNAL_SERVER_ERROR).entity(e.getMessage()).build();
+      }
+    }
+    fileAnswerUtils.setLastEntityId(currentId);
```
Comment (anttiviljakainen) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceSystemRESTService.java]: Can there be problems if batch is too big and the operation times out (does it time out though - maybe database timeout?)? At least the last id is not updated.

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/WorkspaceMaterialFieldAnswerController.java (hunk 1 above)
Comment (pkukkon) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/WorkspaceMaterialFieldAnswerController.java]: Adding existence check and exception if file exists but removal fails.

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java (hunk 2 above)
Comment (pkukkon) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java]: Runtime exceptions in EJBs aren't exactly encouraged but given that these occasions would indicate a serious coding error, I suppose it doesn't matter as much.

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java (hunk 3 above)
Comment (pkukkon) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java]: Yup

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java (hunk 4 above)
Comment (pkukkon) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java]: Changed calling side to catch all exceptions

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java (hunk 5 above)
Comment (pkukkon) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/fieldio/FileAnswerUtils.java]: Yup

**Pull request merged**